components/script/dom/bindings/buffer_source.rs

```diff
@@ -402,44 +407,103 @@ where
-        rooted!(in(*cx) let array_buffer = NewExternalArrayBuffer(
-            *cx,
-            range_size,
-            mapping_slice_ptr as _,
-            Some(free_func),
-            Arc::into_raw(mapping) as _,
-        ));
+    /// Panics if there is any active view
+    pub fn data(&mut self) -> &mut [u8] {
+        // `Arc::get_mut` ensures there are no views
+        Arc::get_mut(&mut self.data).unwrap()
+    }
 
-        HeapBufferSource {
-            buffer_source: BufferSource::ArrayBuffer(Heap::boxed(*array_buffer)),
-            phantom: PhantomData,
+    pub fn clear_views(&mut self) {
+        self.data_views.clear()
+    }
+
+    /// Returns error if requested range is already mapped
+    pub fn view(&mut self, range: Range<usize>) -> Result<&DataView, ()> {
+        if self
+            .data_views
+            .iter()
+            .any(|view| range_overlap(&view.range, &range))
+        {
+            return Err(());
+        }
+        let cx = GlobalScope::get_cx();
+        /// `freeFunc()` must be threadsafe, should be safely callable from any thread
+        /// without causing conflicts, unexpected behavior.
+        unsafe extern "C" fn free_func(_contents: *mut c_void, free_user_data: *mut c_void) {
+            // Clippy warns about "creating a `Arc` from a void raw pointer" here, but suggests
+            // the exact same line to fix it. Doing the cast is tricky because of the use of
+            // a generic type in this parameter.
+            #[allow(clippy::from_raw_with_void_ptr)]
+            drop(Arc::from_raw(free_user_data as *const _));
         }
+        let raw: *mut Box<[u8]> = Arc::into_raw(Arc::clone(&self.data)) as _;
+        rooted!(in(*cx) let object = unsafe {
+            NewExternalArrayBuffer(
+                *cx,
+                range.end - range.start,
+                // SAFETY: This is safe because we have checked there is no overlapping view
+                (*raw)[range.clone()].as_mut_ptr() as _,
+                Some(free_func),
+                raw as _,
+            )
+        });
+        self.data_views.push(DataView {
+            range,
+            buffer: HeapArrayBuffer::from(*object).unwrap(),
+        });
+        Ok(self.data_views.last().unwrap())
+    }
+}
+
+#[derive(JSTraceable, MallocSizeOf)]
+pub struct DataView {
+    #[no_trace]
+    range: Range<usize>,
+    #[ignore_malloc_size_of = "defined in mozjs"]
+    buffer: HeapArrayBuffer,
+}
+
+impl DataView {
+    pub fn array_buffer(&self) -> ArrayBuffer {
+        unsafe { ArrayBuffer::from(self.buffer.underlying_object().get()).unwrap() }
+    }
+}
+
+impl Drop for DataView {
+    #[allow(unsafe_code)]
+    fn drop(&mut self) {
+        let cx = GlobalScope::get_cx();
+        assert!(unsafe {
+            js::jsapi::DetachArrayBuffer(*cx, self.buffer.underlying_object().handle())
```

Comment:
nit: import at top

Reply:
It is already imported (although wrapped version) which we cannot use here as we have raw handle (actually we could do `Handle::from_raw()`, but I think that would be more work for less profit).